src/python_testing/TC_pics_checker.py

```diff
@@ -0,0 +1,152 @@
+        self.xml_clusters, self.problems = build_xml_clusters()
+
+    def _check_and_record_errors(self, location, required, pics):
+        if required and not self.check_pics(pics):
+            self.record_error("PICS check", location=location, problem=f"Required PICS {pics} not found in pics list")
+            self.success = False
+        elif not required and self.check_pics(pics):
+            self.record_error("PICS check", location=location, problem=f"PICS {pics} found in PICS list, but not on device")
+            self.success = False
+
+    def _add_pics_for_lists(self, cluster_id: int, global_attribute: GlobalAttributeIds) -> None:
+        if global_attribute == GlobalAttributeIds.ATTRIBUTE_LIST_ID:
+            cluster_object = Clusters.ClusterObjects.ALL_ATTRIBUTES
+            pics_function = attribute_pics
+
+        elif global_attribute == GlobalAttributeIds.ACCEPTED_COMMAND_LIST_ID:
+            cluster_object = Clusters.ClusterObjects.ALL_ACCEPTED_COMMANDS
+            pics_function = accepted_cmd_pics
+
+        elif global_attribute == GlobalAttributeIds.GENERATED_COMMAND_LIST_ID:
+            cluster_object = Clusters.ClusterObjects.ALL_GENERATED_COMMANDS
+            pics_function = generated_cmd_pics
+
+        else:
+            asserts.fail("add_pics_for_list function called for non-list attribute")
+
+        # This cluster does not have any of this element type
+        if cluster_id not in cluster_object.keys():
+            return
+
+        for element_id in cluster_object[cluster_id]:
+            if element_id > 0xF000:
+                # No pics for global elements
+                continue
+            pics = pics_function(self.xml_clusters[cluster_id].pics, element_id)
+
+            if cluster_id not in self.endpoint.keys():
+                # This cluster is not on this endpoint
+                required = False
+            elif element_id in self.endpoint[cluster_id][global_attribute]:
+                # Cluster and element are on the endpoint
+                required = True
+            else:
+                # Cluster is on the endpoint but the element is not in the list
+                required = False
+
+            if global_attribute == GlobalAttributeIds.ATTRIBUTE_LIST_ID:
+                location = AttributePathLocation(endpoint_id=self.endpoint_id, cluster_id=cluster_id, attribute_id=element_id)
+            else:
+                location = CommandPathLocation(endpoint_id=self.endpoint_id, cluster_id=cluster_id, command_id=element_id)
+
+            self._check_and_record_errors(location, required, pics)
+
+    def test_TC_pics_checker(self):
+        self.endpoint_id = self.matter_test_config.endpoint
+        self.endpoint = self.endpoints_tlv[self.endpoint_id]
+        self.success = True
+
+        for cluster_id, cluster in Clusters.ClusterObjects.ALL_CLUSTERS.items():
+            if cluster_id not in self.xml_clusters or self.xml_clusters[cluster_id].pics is None:
```

Comment:
Can we add a small comment to why the XML stuff is used?

Reply:
added.